File: src/switch_tracker/services/inspect.py
"""Worker: the click-to-pick selector tool, in a REAL, VISIBLE browser window.

The one capability that cannot live inside the dashboard's own tab -- a human
has to click on the store's actual page.  It is still launched exactly like
every other action (the exe re-invoking itself), so it is not a special case
in the architecture, only in being visible.

Clicks are resolved through ``elementsFromPoint`` rather than the click event's
own target: many product cards wrap the whole tile in one absolutely-positioned
overlay link, so a plain click always hits that overlay and never the title or
price text visually underneath it.  Reading the entire element stack at the
click point recovers the real element.
"""

from __future__ import annotations

import asyncio
from typing import Any

from switch_tracker.adapters.browser import overrides as profile_overrides
from switch_tracker.adapters.browser.provider import BrowserProvider
from switch_tracker.config import overrides
from switch_tracker.core import db
from switch_tracker.events.writer import EventWriter

STEPS = (
    ("card", "Alt+click the WHOLE PRODUCT CARD -- the box around one listing."),
    ("title", "Alt+click that card's TITLE text."),
    ("price", "Alt+click that card's PRICE text."),
    ("link", "Alt+click that card's LINK (often the title itself, or the image)."),
)

_PICKER = """
() => {
  let hovered = null;

  const candidatesFor = (el) => {
    const tag = el.tagName.toLowerCase();
    const tried = new Map();
    const add = (css) => {
      if (tried.has(css)) return;
      try { tried.set(css, { css, matches: document.querySelectorAll(css).length }); }
      catch { /* not valid CSS here */ }
    };
    // Ordered stable-first: an id beats a data attribute beats a class name.
    if (el.id) add('#' + CSS.escape(el.id));
    for (const attr of ['data-testid','data-cy','data-component-type','data-id',
                        'data-product-id','data-asin','data-sku','data-pid']) {
      const v = el.getAttribute(attr);
      if (v !== null) { add(`${tag}[${attr}='${CSS.escape(v)}']`); add(`[${attr}]`); }
    }
    const classes = [...el.classList].map((c) => CSS.escape(c));
    if (classes.length) add(`${tag}.${classes.join('.')}`);
    add(tag);
    return [...tried.values()];
  };

  document.addEventListener('mouseover', (e) => {
    if (hovered) hovered.style.outline = '';
    hovered = e.target;
    hovered.style.outline = '2px solid #ff2d78';
  }, true);

  document.addEventListener('click', (e) => {
    if (!e.altKey) return;   // plain clicks still dismiss banners and navigate
    e.preventDefault();
    e.stopPropagation();

    const stack = document.elementsFromPoint(e.clientX, e.clientY)
      .filter((el) => el.tagName !== 'HTML' && el.tagName !== 'BODY')
      .slice(0, 8);

    const seen = new Set();
    const candidates = [];
    for (const el of stack) {
      for (const c of candidatesFor(el)) {
        if (seen.has(c.css)) continue;
        seen.add(c.css);
        candidates.push(c);
      }
    }
    window.__reportPick({ candidates });
  }, true);
}
"""


async def run(run_id: int, store_id: str) -> int:
    conn = db.connect()
    db.ensure_schema(conn)
    events = EventWriter(conn, run_id)
    events.run_started(1)

    store = next((s for s in overrides.active_stores() if s.id == store_id), None)
    if store is None or not store.search_urls:
        events.warning(f"{store_id} has no page to open.")
        events.run_finished()
        conn.execute("UPDATE run SET finished_at = datetime('now') WHERE id = ?", (run_id,))
        return 1

    provider = BrowserProvider(headless=False)
    step_index = 0
    card_selector: str | None = None
    finished = asyncio.Event()

    try:
        context = await provider.new_context()
        page = await context.new_page()

        async def coverage(css: str) -> float:
            """How many of the sampled cards actually contain this selector."""
            if card_selector is None:
                return 0.0
            return float(
                await page.evaluate(
                    """([cardSel, css]) => {
                        const cards = Array.from(document.querySelectorAll(cardSel)).slice(0, 8);
                        if (!cards.length) return 0;
                        let hits = 0;
                        for (const card of cards) {
                          try { if (card.querySelector(css)) hits++; } catch { /* skip */ }
                        }
                        return hits / cards.length;
                    }""",
                    [card_selector, css],
                )
            )

        async def report_pick(pick: dict[str, Any]) -> None:
            nonlocal step_index, card_selector
            if step_index >= len(STEPS):
                step_index = 0
            field, _ = STEPS[step_index]
            candidates = pick.get("candidates") or []

            if field == "card":
                # A card selector has to REPEAT. Candidates arrive stable-first,
                # so the first that matches a sane number of times wins.
                winner = next((c for c in candidates if 2 <= c["matches"] <= 500), None)
                if winner is None:
                    best = candidates[0]["matches"] if candidates else 0
                    events.warning(
                        f"Nothing there repeats across the page (best: {best} matches) -- "
                        "try the tile's outer box."
                    )
                    return
                card_selector = winner["css"]
                profile_overrides.add_selector(store_id, "card", winner["css"])
                events.warning(f"card: {winner['css']} ({winner['matches']} on this page) -- saved.")
            else:
                scored = [(await coverage(c["css"]), c["css"]) for c in candidates]
                scored.sort(reverse=True)
                if not scored or scored[0][0] == 0:
                    events.warning("That is not inside the saved card -- click inside a tile.")
                    return
                score, css = scored[0]
                profile_overrides.add_selector(store_id, field, css)
                events.warning(f"{field}: {css} (in {round(score * 100)}% of sampled cards) -- saved.")

            step_index += 1
            if step_index >= len(STEPS):
                events.warning(
                    "All four saved. Run 'Collect prices' to try them, or keep Alt+clicking "
                    "to add fallbacks for listings that look different."
                )
                step_index = 0
            else:
                events.warning(f"{step_index + 1}/4 -- {STEPS[step_index][1]}")

        await page.expose_function("__reportPick", report_pick)
        await page.add_init_script(_PICKER)
        page.on("close", lambda _: finished.set())

        await page.goto(store.search_urls[0].replace("{p}", "1"))
        events.warning(f"Opened {store.name}. 1/4 -- {STEPS[0][1]}")
        events.warning("Plain clicks still work normally. Close the window when you are done.")

        await finished.wait()
    finally:
        await provider.close()

    events.run_finished()
    conn.execute("UPDATE run SET finished_at = datetime('now') WHERE id = ?", (run_id,))
    return 0

File: src/cli/inspect.ts
import { chromium } from "playwright";
import { activeStores } from "../config/overrides.ts";
import { addProfileSelector } from "../adapters/browser.ts";
const STORES = activeStores();

/**
 * Opens a hard store's search page and guides you through building its
 * PROFILES entry by clicking, with zero manual editing afterward.
 *
 *   bun run src/cli/inspect.ts playasia
 *
 * Alt+click the card, then its title, price and link, in that order — each
 * click is scored automatically (a card selector must repeat across the
 * page; title/price/link are scored by how many sampled cards actually
 * contain them) and the winner is written straight to profiles.local.json,
 * which browser.ts merges over its built-in guesses at fetch time. Nothing
 * needs pasting into browser.ts by hand.
 *
 * Playwright Inspector's own "pick locator" isn't used here on purpose — it
 * produces `getByRole(...)`-style locator code for test scripts, not plain
 * CSS, and there's no way to read what you picked there back into this
 * script. Plain clicks (no Alt) still behave normally, so cookie banners and
 * "continue" buttons can still be dismissed.
 *
 * Clicks are resolved via `elementsFromPoint`, not `event.target` — a lot of
 * product cards wrap the whole tile in one invisible/absolute-positioned
 * overlay link for click-to-navigate, so a plain click always hits that
 * overlay, never the title/price text visually underneath it. Reading the
 * whole element stack at the click point recovers the real element even
 * when a click alone couldn't reach it.
 */
const id = process.argv[2];
const store = STORES.find((s) => s.id === id);

if (!store?.searchUrls?.[0]) {
  console.log("usage: bun run src/cli/inspect.ts <storeId>");
  console.log("browser stores: " + STORES.filter((s) => s.kind === "BROWSER").map((s) => s.id).join(", "));
  process.exit(1);
}

type Field = "cardSelectors" | "titleSelectors" | "priceSelectors" | "linkSelectors";

const STEPS: { field: Field; prompt: string }[] = [
  { field: "cardSelectors", prompt: "Alt+click the WHOLE PRODUCT CARD — the box around one listing." },
  { field: "titleSelectors", prompt: "Alt+click that card's TITLE text." },
  { field: "priceSelectors", prompt: "Alt+click that card's PRICE text." },
  { field: "linkSelectors", prompt: "Alt+click that card's LINK (often the title itself, or the image)." },
];

interface Candidate {
  css: string;
  matches: number;
}

interface Pick {
  tag: string;
  text: string;
  candidates: Candidate[];
}

let stepIndex = 0;
let cardSelector: string | undefined;

function announce(): void {
  const step = STEPS[stepIndex];
  if (step) {
    console.log(`\n${stepIndex + 1}/4 — ${step.prompt}`);
  } else {
    console.log(
      `\nAll four saved to profiles.local.json for "${store.id}". Run "bun run collect" to try it, ` +
        `or keep Alt+clicking — another round starts over from the card, useful for a second example ` +
        `if some listings look different (e.g. one on sale).`,
    );
    stepIndex = 0; // let another round of examples add fallback candidates
  }
}

const browser = await chromium.launch({ headless: false });
const page = await browser.newPage({ locale: "en-IN", timezoneId: "Asia/Kolkata" });

async function coverageWithinCards(css: string): Promise<number> {
  if (!cardSelector) return 0;
  return page.evaluate(
    ({ cardSel, css }) => {
      const cards = Array.from(document.querySelectorAll(cardSel)).slice(0, 8);
      if (cards.length === 0) return 0;
      let hits = 0;
      for (const card of cards) {
        try {
          if (card.querySelector(css)) hits++;
        } catch {
          // invalid selector in this DOM context — treat as a miss
        }
      }
      return hits / cards.length;
    },
    { cardSel: cardSelector, css },
  );
}

await page.exposeFunction("__reportPick", async (pick: Pick) => {
  const step = STEPS[stepIndex];
  if (!step) return;

  if (step.field === "cardSelectors") {
    // A card selector has to repeat — candidates arrive ordered stable-first
    // (id, then value-specific attribute, then bare attribute, then class,
    // then tag), so the first one that repeats a sane number of times wins.
    const winner = pick.candidates.find((c) => c.matches >= 2 && c.matches <= 500);
    if (!winner) {
      const best = pick.candidates[0]?.matches ?? 0;
      console.log(`   ✗ nothing there repeats across the page (best match count: ${best}) — try the tile's outer box instead.`);
      return;
    }
    cardSelector = winner.css;
    addProfileSelector(store.id, "cardSelectors", winner.css);
    console.log(`   ✓ card: ${winner.css}  (${winner.matches} listings on this page) — saved.`);
  } else {
    let best: { css: string; coverage: number } | undefined;
    for (const c of pick.candidates) {
      const coverage = await coverageWithinCards(c.css);
      if (!best || coverage > best.coverage) best = { css: c.css, coverage };
    }
    if (!best || best.coverage === 0) {
      console.log(`   ✗ that doesn't appear inside the saved card selector — click something inside one of the highlighted tiles.`);
      return;
    }
    addProfileSelector(store.id, step.field, best.css);
    console.log(`   ✓ ${step.field}: ${best.css}  (found in ${Math.round(best.coverage * 100)}% of sampled cards) — saved.`);
  }

  stepIndex++;
  announce();
});

// Runs in the browser, not Node — re-injected on every navigation.
await page.addInitScript(() => {
  let hovered: HTMLElement | null = null;

  function candidatesFor(el: Element): { css: string; matches: number }[] {
    const tag = el.tagName.toLowerCase();
    const tried = new Map<string, { css: string; matches: number }>();

    const add = (css: string) => {
      if (tried.has(css)) return;
      try {
        tried.set(css, { css, matches: document.querySelectorAll(css).length });
      } catch {
        // not valid CSS in this context — skip
      }
    };

    if (el.id) add(`#${CSS.escape(el.id)}`);
    for (const attr of ["data-testid", "data-cy", "data-component-type", "data-id", "data-product-id", "data-asin", "data-sku", "data-pid"]) {
      const v = el.getAttribute(attr);
      if (v !== null) {
        add(`${tag}[${attr}='${CSS.escape(v)}']`);
        add(`[${attr}]`);
      }
    }
    const classes = [...el.classList].map((c) => CSS.escape(c));
    if (classes.length) add(`${tag}.${classes.join(".")}`);
    add(tag);
    return [...tried.values()];
  }

  document.addEventListener(
    "mouseover",
    (e) => {
      if (hovered) hovered.style.outline = "";
      hovered = e.target as HTMLElement;
      hovered.style.outline = "2px solid #ff2d78";
    },
    true,
  );

  document.addEventListener(
    "click",
    (e) => {
      if (!e.altKey) return; // plain clicks still dismiss banners, navigate, etc.
      e.preventDefault();
      e.stopPropagation();

      // A card's title/price often sit UNDER a full-tile overlay <a> used for
      // click-to-navigate — e.target on a click there is always that overlay,
      // never the text visually underneath it. elementsFromPoint returns the
      // whole stack at that point, front-to-back, so the real title/price
      // element is still in there even when it isn't what a plain click would
      // have hit.
      const stack = document
        .elementsFromPoint(e.clientX, e.clientY)
        .filter((el) => el.tagName !== "HTML" && el.tagName !== "BODY")
        .slice(0, 8);
      const top = stack[0] ?? (e.target as Element);

      const seen = new Set<string>();
      const candidates: { css: string; matches: number }[] = [];
      for (const el of stack) {
        for (const c of candidatesFor(el)) {
          if (seen.has(c.css)) continue;
          seen.add(c.css);
          candidates.push(c);
        }
      }

      (window as unknown as { __reportPick: (p: unknown) => void }).__reportPick({
        tag: top.tagName.toLowerCase(),
        text: (top.textContent ?? "").trim().replace(/\s+/g, " "),
        candidates,
      });
    },
    true,
  );
});

await page.goto(store.searchUrls[0].replace("{p}", "1"));

console.log(`\nOpened ${store.name}.`);
announce();
console.log("\nPlain clicks still behave normally (dismiss popups, scroll, etc). Close the browser when done.\n");

await page.waitForEvent("close", { timeout: 0 }).catch(() => undefined);
await browser.close();
